File: sudoku_game.py
import tkinter as tk
from tkinter import messagebox
import random
import copy

class SudokuGame:
    def __init__(self, master):
        self.master = master
        self.master.title("数独游戏")
        self.master.geometry("600x700")
        self.master.configure(bg='#f0f0f0')
        
        # 游戏状态
        self.board = [[0 for _ in range(9)] for _ in range(9)]
        self.solution = [[0 for _ in range(9)] for _ in range(9)]
        self.original_board = [[0 for _ in range(9)] for _ in range(9)]
        self.selected_cell = None
        self.cells = []
        
        # 创建UI
        self.create_widgets()
        self.new_game()
        
    def create_widgets(self):
        # 标题
        title_label = tk.Label(self.master, text="数独游戏", 
                              font=("Arial", 24, "bold"), bg='#f0f0f0')
        title_label.pack(pady=10)
        
        # 游戏板框架
        self.game_frame = tk.Frame(self.master, bg='black', bd=3, relief=tk.RAISED)
        self.game_frame.pack(pady=10)
        
        # 创建9x9的网格
        for i in range(9):
            row = []
            for j in range(9):
                # 计算边框宽度
                border_top = 3 if i % 3 == 0 else 1
                border_left = 3 if j % 3 == 0 else 1
                border_bottom = 3 if i == 8 else 1
                border_right = 3 if j == 8 else 1
                
                cell_frame = tk.Frame(self.game_frame, bg='black',
                                    highlightthickness=0)
                cell_frame.grid(row=i, column=j, padx=(border_left, border_right),
                              pady=(border_top, border_bottom))
                
                cell = tk.Label(cell_frame, text="", width=3, height=2,
                              font=("Arial", 18, "bold"), bg='white',
                              relief=tk.FLAT, cursor="hand2")
                cell.pack()
                cell.bind("<Button-1>", lambda e, row=i, col=j: self.cell_clicked(row, col))
                
                row.append(cell)
            self.cells.append(row)
        
        # 控制按钮框架
        control_frame = tk.Frame(self.master, bg='#f0f0f0')
        control_frame.pack(pady=10)
        
        # 新游戏按钮
        new_game_btn = tk.Button(control_frame, text="新游戏", 
                               command=self.new_game,
                               font=("Arial", 12), bg='#4CAF50', fg='white',
                               padx=20, pady=10, relief=tk.RAISED)
        new_game_btn.pack(side=tk.LEFT, padx=5)
        
        # 检查答案按钮
        check_btn = tk.Button(control_frame, text="检查答案", 
                            command=self.check_solution,
                            font=("Arial", 12), bg='#2196F3', fg='white',
                            padx=20, pady=10, relief=tk.RAISED)
        check_btn.pack(side=tk.LEFT, padx=5)
        
        # 提示按钮
        hint_btn = tk.Button(control_frame, text="提示", 
                           command=self.give_hint,
                           font=("Arial", 12), bg='#FF9800', fg='white',
                           padx=20, pady=10, relief=tk.RAISED)
        hint_btn.pack(side=tk.LEFT, padx=5)
        
        # 数字按钮框架
        number_frame = tk.Frame(self.master, bg='#f0f0f0')
        number_frame.pack(pady=10)
        
        # 数字按钮1-9
        for num in range(1, 10):
            btn = tk.Button(number_frame, text=str(num), 
                          command=lambda n=num: self.number_clicked(n),
                          font=("Arial", 14, "bold"), bg='#9C27B0', fg='white',
                          width=4, height=2, relief=tk.RAISED)
            btn.grid(row=0, column=num-1, padx=2, pady=2)
        
        # 清除按钮
        clear_btn = tk.Button(number_frame, text="清除", 
                            command=self.clear_cell,
                            font=("Arial", 12), bg='#f44336', fg='white',
                            padx=15, pady=10, relief=tk.RAISED)
        clear_btn.grid(row=1, column=4, padx=2, pady=2)
        
        # 绑定键盘事件
        self.master.bind("<Key>", self.key_pressed)
        
    def generate_sudoku(self):
        # 生成完整的数独解决方案
        board = [[0 for _ in range(9)] for _ in range(9)]
        self.fill_board(board)
        self.solution = copy.deepcopy(board)
        
        # 移除一些数字创建谜题
        cells_to_remove = random.randint(40, 55)
        for _ in range(cells_to_remove):
            row = random.randint(0, 8)
            col = random.randint(0, 8)
            board[row][col] = 0
        
        self.board = copy.deepcopy(board)
        self.original_board = copy.deepcopy(board)
        
    def fill_board(self, board):
        # 使用回溯算法填充数独板
        empty = self.find_empty(board)
        if not empty:
            return True
        
        row, col = empty
        numbers = list(range(1, 10))
        random.shuffle(numbers)
        
        for num in numbers:
            if self.is_valid(board, num, row, col):
                board[row][col] = num
                
                if self.fill_board(board):
                    return True
                
                board[row][col] = 0
        
        return False
    
    def find_empty(self, board):
        for i in range(9):
            for j in range(9):
                if board[i][j] == 0:
                    return (i, j)
        return None
    
    def is_valid(self, board, num, row, col):
        # 检查行
        for j in range(9):
            if board[row][j] == num:
                return False
        
        # 检查列
        for i in range(9):
            if board[i][col] == num:
                return False
        
        # 检查3x3宫格
        box_row = (row // 3) * 3
        box_col = (col // 3) * 3
        for i in range(box_row, box_row + 3):
            for j in range(box_col, box_col + 3):
                if board[i][j] == num:
                    return False
        
        return True
    
    def new_game(self):
        self.generate_sudoku()
        self.update_display()
        self.selected_cell = None
        
    def update_display(self):
        for i in range(9):
            for j in range(9):
                value = self.board[i][j]
                cell = self.cells[i][j]
                
                if value == 0:
                    cell.config(text="", bg='white')
                else:
                    if self.original_board[i][j] != 0:
                        cell.config(text=str(value), bg='#e0e0e0', fg='black')
                    else:
                        cell.config(text=str(value), bg='white', fg='blue')
                
                # 高亮选中的单元格
                if self.selected_cell == (i, j):
                    cell.config(bg='#ffeb3b')
                elif self.selected_cell and self.is_same_group(self.selected_cell, (i, j)):
                    cell.config(bg='#fff9c4')
    
    def is_same_group(self, cell1, cell2):
        row1, col1 = cell1
        row2, col2 = cell2
        
        # 同一行
        if row1 == row2:
            return True
        # 同一列
        if col1 == col2:
            return True
        # 同一宫格
        if (row1 // 3) == (row2 // 3) and (col1 // 3) == (col2 // 3):
            return True
        
        return False
    
    def cell_clicked(self, row, col):
        self.selected_cell = (row, col)
        self.update_display()
    
    def number_clicked(self, num):
        if self.selected_cell:
            row, col = self.selected_cell
            if self.original_board[row][col] == 0:  # 只能修改非原始数字
                self.board[row][col] = num
                self.update_display()
    
    def clear_cell(self):
        if self.selected_cell:
            row, col = self.selected_cell
            if self.original_board[row][col] == 0:  # 只能清除非原始数字
                self.board[row][col] = 0
                self.update_display()
    
    def key_pressed(self, event):
        if self.selected_cell:
            row, col = self.selected_cell
            if self.original_board[row][col] == 0:  # 只能修改非原始数字
                if event.char.isdigit() and event.char != '0':
                    self.board[row][col] = int(event.char)
                    self.update_display()
                elif event.keysym in ['BackSpace', 'Delete', 'Clear']:
                    self.board[row][col] = 0
                    self.update_display()
    
    def check_solution(self):
        # 检查当前解答是否正确
        for i in range(9):
            for j in range(9):
                if self.board[i][j] == 0:
                    messagebox.showwarning("未完成", "数独还没有完成！")
                    return
        
        # 检查是否与解决方案匹配
        for i in range(9):
            for j in range(9):
                if self.board[i][j] != self.solution[i][j]:
                    messagebox.showerror("错误", "答案有误，请继续努力！")
                    return
        
        messagebox.showinfo("恭喜", "恭喜你！数独解答正确！")
    
    def give_hint(self):
        # 找到一个空单元格并填入正确答案
        empty_cells = []
        for i in range(9):
            for j in range(9):
                if self.board[i][j] == 0:
                    empty_cells.append((i, j))
        
        if empty_cells:
            row, col = random.choice(empty_cells)
            self.board[row][col] = self.solution[row][col]
            self.selected_cell = (row, col)
            self.update_display()
        else:
            messagebox.showinfo("提示", "没有空单元格需要提示！")

def main():
    root = tk.Tk()
    game = SudokuGame(root)
    root.mainloop()

if __name__ == "__main__":
    main()

File: sudoku_game.cpp
#include <iostream>
#include <vector>
#include <random>
#include <ctime>
#include <algorithm>
#include <iomanip>
#include <limits>

using namespace std;

class SudokuGame {
private:
    vector<vector<int>> board;
    vector<vector<int>> solution;
    vector<vector<int>> original_board;
    int selected_row;
    int selected_col;
    
public:
    SudokuGame() : selected_row(-1), selected_col(-1) {
        board = vector<vector<int>>(9, vector<int>(9, 0));
        solution = vector<vector<int>>(9, vector<int>(9, 0));
        original_board = vector<vector<int>>(9, vector<int>(9, 0));
    }
    
    void generate_sudoku() {
        // 生成完整的数独解决方案
        fill_board(solution);
        
        // 复制解决方案到游戏板
        board = solution;
        original_board = solution;
        
        // 移除一些数字创建谜题
        int cells_to_remove = 40 + rand() % 16; // 40-55个空格
        for (int i = 0; i < cells_to_remove; i++) {
            int row = rand() % 9;
            int col = rand() % 9;
            board[row][col] = 0;
            original_board[row][col] = 0;
        }
    }
    
    bool fill_board(vector<vector<int>>& board) {
        for (int row = 0; row < 9; row++) {
            for (int col = 0; col < 9; col++) {
                if (board[row][col] == 0) {
                    vector<int> numbers = {1, 2, 3, 4, 5, 6, 7, 8, 9};
                    random_shuffle(numbers.begin(), numbers.end());
                    
                    for (int num : numbers) {
                        if (is_valid(board, num, row, col)) {
                            board[row][col] = num;
                            
                            if (fill_board(board)) {
                                return true;
                            }
                            
                            board[row][col] = 0;
                        }
                    }
                    return false;
                }
            }
        }
        return true;
    }
    
    bool is_valid(const vector<vector<int>>& board, int num, int row, int col) {
        // 检查行
        for (int j = 0; j < 9; j++) {
            if (board[row][j] == num) {
                return false;
            }
        }
        
        // 检查列
        for (int i = 0; i < 9; i++) {
            if (board[i][col] == num) {
                return false;
            }
        }
        
        // 检查3x3宫格
        int box_row = (row / 3) * 3;
        int box_col = (col / 3) * 3;
        for (int i = box_row; i < box_row + 3; i++) {
            for (int j = box_col; j < box_col + 3; j++) {
                if (board[i][j] == num) {
                    return false;
                }
            }
        }
        
        return true;
    }
    
    void display_board() {
        cout << "\n   ";
        for (int i = 0; i < 9; i++) {
            cout << " " << i + 1 << "  ";
        }
        cout << "\n   +---+---+---+---+---+---+---+---+---+\n";
        
        for (int i = 0; i < 9; i++) {
            cout << i + 1 << " |";
            for (int j = 0; j < 9; j++) {
                if (get_selected_row() == i && get_selected_col() == j) {
                    cout << "[" << setw(2) << (board[i][j] ? to_string(board[i][j]) : " ") << "]";
                } else if (original_board[i][j] != 0) {
                    cout << " " << setw(2) << board[i][j] << " ";
                } else {
                    cout << " " << setw(2) << (board[i][j] ? to_string(board[i][j]) : " ") << " ";
                }
                
                if (j % 3 == 2 && j < 8) {
                    cout << "|";
                }
            }
            cout << "|\n";
            
            if (i % 3 == 2 && i < 8) {
                cout << "   +---+---+---+---+---+---+---+---+---+\n";
            }
        }
        cout << "   +---+---+---+---+---+---+---+---+---+\n";
    }
    
    void select_cell(int row, int col) {
        if (row >= 0 && row < 9 && col >= 0 && col < 9) {
            set_selected_row(row);
            set_selected_col(col);
        }
    }
    
    int get_selected_row() const { return selected_row; }
    int get_selected_col() const { return selected_col; }
    void set_selected_row(int row) { selected_row = row; }
    void set_selected_col(int col) { selected_col = col; }
    
    bool place_number(int num) {
        if (selected_row == -1 || selected_col == -1) {
            cout << "请先选择一个单元格！\n";
            return false;
        }
        
        if (original_board[selected_row][selected_col] != 0) {
            cout << "这个单元格是固定的，不能修改！\n";
            return false;
        }
        
        if (num < 1 || num > 9) {
            cout << "数字必须在1-9之间！\n";
            return false;
        }
        
        board[selected_row][selected_col] = num;
        return true;
    }
    
    void clear_cell() {
        if (selected_row == -1 || selected_col == -1) {
            cout << "请先选择一个单元格！\n";
            return;
        }
        
        if (original_board[selected_row][selected_col] != 0) {
            cout << "这个单元格是固定的，不能清除！\n";
            return;
        }
        
        board[selected_row][selected_col] = 0;
    }
    
    bool is_complete() {
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == 0) {
                    return false;
                }
            }
        }
        return true;
    }
    
    bool is_correct() {
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] != solution[i][j]) {
                    return false;
                }
            }
        }
        return true;
    }
    
    void give_hint() {
        vector<pair<int, int>> empty_cells;
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == 0) {
                    empty_cells.push_back({i, j});
                }
            }
        }
        
        if (empty_cells.empty()) {
            cout << "没有空单元格需要提示！\n";
            return;
        }
        
        int index = rand() % empty_cells.size();
        int row = empty_cells[index].first;
        int col = empty_cells[index].second;
        
        board[row][col] = solution[row][col];
        selected_row = row;
        selected_col = col;
        
        cout << "提示：单元格 (" << row + 1 << "," << col + 1 << ") 的答案是 " << solution[row][col] << "\n";
    }
    
    void new_game() {
        generate_sudoku();
        selected_row = -1;
        selected_col = -1;
        cout << "新游戏开始！\n";
    }
    
    void show_instructions() {
        cout << "\n=== 数独游戏说明 ===\n";
        cout << "1. 输入行号和列号选择单元格 (1-9)\n";
        cout << "2. 输入数字1-9在选中单元格放置数字\n";
        cout << "3. 输入0清除选中单元格\n";
        cout << "4. 输入'c'清除当前单元格\n";
        cout << "5. 输入'h'获取提示\n";
        cout << "6. 输入'check'检查答案\n";
        cout << "7. 输入'new'开始新游戏\n";
        cout << "8. 输入'quit'退出游戏\n";
        cout << "==================\n";
    }
};

int main() {
    srand(time(0));
    
    SudokuGame game;
    game.new_game();
    game.show_instructions();
    
    while (true) {
        game.display_board();
        
        cout << "\n当前选择：";
        if (game.selected_row != -1 && game.selected_col != -1) {
            cout << "(" << game.selected_row + 1 << "," << game.selected_col + 1 << ")";
        } else {
            cout << "无";
        }
        cout << "\n请输入命令：";
        
        string input;
        cin >> input;
        
        if (input == "quit") {
            cout << "游戏结束，再见！\n";
            break;
        }
        else if (input == "new") {
            game.new_game();
            continue;
        }
        else if (input == "check") {
            if (game.is_complete()) {
                if (game.is_correct()) {
                    cout << "恭喜！数独解答正确！\n";
                } else {
                    cout << "答案有误，请继续努力！\n";
                }
            } else {
                cout << "数独还没有完成！\n";
            }
            continue;
        }
        else if (input == "h") {
            game.give_hint();
            continue;
        }
        else if (input == "c") {
            game.clear_cell();
            continue;
        }
        else if (input == "help") {
            game.show_instructions();
            continue;
        }
        
        // 尝试解析坐标或数字
        try {
            if (input.length() == 1 && input[0] >= '1' && input[0] <= '9') {
                // 选择行
                int row = stoi(input) - 1;
                cout << "请输入列号 (1-9)：";
                string col_input;
                cin >> col_input;
                int col = stoi(col_input) - 1;
                game.select_cell(row, col);
            }
            else if (input.length() == 2 && input[0] >= '1' && input[0] <= '9' && 
                     input[1] >= '1' && input[1] <= '9') {
                // 选择单元格
                int row = stoi(input.substr(0, 1)) - 1;
                int col = stoi(input.substr(1, 2)) - 1;
                game.select_cell(row, col);
            }
            else if (input.length() == 3 && input[0] >= '1' && input[0] <= '9' && 
                     input[1] >= '1' && input[1] <= '9' && input[2] >= '1' && input[2] <= '9') {
                // 选择单元格并放置数字
                int row = stoi(input.substr(0, 1)) - 1;
                int col = stoi(input.substr(1, 2)) - 1;
                int num = stoi(input.substr(2, 3));
                game.select_cell(row, col);
                game.place_number(num);
            }
            else {
                cout << "无效的输入！请输入有效的坐标或命令。\n";
            }
        } catch (...) {
            cout << "输入格式错误！请重新输入。\n";
            cin.clear();
            cin.ignore(numeric_limits<streamsize>::max(), '\n');
        }
    }
    
    return 0;
}